### ISSUE TITLE
Rewrite optimization_train_loop.py as thin wrapper over src/training/ modules

### ISSUE THREAD

**Issue by zeinali72** [zeinali72/swe-pinn#91]
## Context

`optimisation/optimization_train_loop.py` (423 lines) is a **frozen copy** of training logic from an older version of `src/train.py`. The `src/` modules have since been refactored into a modular architecture (`src/training/step.py`, `src/training/epoch.py`, `src/training/optimizer.py`, `src/training/setup.py`), and the experiment scripts in `experiments/` now use these clean modules. The HPO training loop has **not kept pace**, resulting in divergent behaviour between HPO and production.

Full analysis: `optimisation/ANALYSIS_AND_CHANGES.md` (sections 3 and 4).

## Related Issues

- **Depends on**: #88 (extract experiment factories — needed so we can import `compute_losses` and `make_generate_epoch_data`)
- **Subsumes/resolves when complete**:
  - Stale `src.data` imports (`get_sample_count`, `get_batches_tensor`)
  - Duplicate building mask computation (lines 55+63)
  - Missing `min_depth` clipping in validation
  - Optimizer mismatch (`piecewise_constant_schedule` vs `reduce_on_plateau`)
- **Parallel-safe** with: #89 (config cleanup), #90 (archive cleanup)
- **Enables**: experiment dispatch registry, loss term tracking, patience-based early stopping, data-driven HPO, HPO unit tests (all filed separately)

## Problem Summary

| Aspect | Current HPO | Production (`src/training/`) |
|--------|------------|------------------------------|
| Optimizer | `chain(clip(1.0), adam(piecewise))` | `chain(clip(cfg), adam(piecewise), reduce_on_plateau)` |
| Loss function | Inlined 60-line closure in `train_step_trial()` | `compute_losses()` callback via `make_scan_body()` |
| Weight handling | `k.replace('_weight','')` + `k != 'data'` | `extract_loss_weights()` + `get_active_loss_weights(data_free=...)` |
| Batch calculation | Manual `max(n//bs for n in counts)` | `calculate_num_batches(batch_size, sizes, data, data_free)` |
| Boundary segments | `max(5, n//4)` | `get_boundary_segment_count(cfg, n)` |
| Data sampling | Raw `sample_domain()` + `get_batches_tensor()` | `sample_and_batch()` from `src/training/epoch.py` |
| Scan body | Returns `(carry, None)` — discards losses | Returns `(carry, (terms, total))` via `make_scan_body()` |
| Validation | No `min_depth`, only `nse_h` | `_apply_min_depth()`, `nse_h/hu/hv`, `rmse_h/hu/hv` |
| Model init | `importlib.import_module` + `getattr` | `init_model_from_config(cfg)` |

## Proposed Change

Replace the entire 423-line file with a ~100-150 line thin wrapper that:

### 1. Uses `src/training/` for all training logic

```python
from src.training import (
    create_optimizer,
    extract_loss_weights, get_active_loss_weights,
    calculate_num_batches, get_sampling_count_from_config,
    get_boundary_segment_count,
    make_scan_body, train_step_jitted,
    sample_and_batch, maybe_batch_data, empty_batch,
)
from src.models import init_model_from_config
from src.utils import nse
from src.predict.predictor import _apply_min_depth
```

### 2. Imports experiment-specific logic from experiments/

```python
from experiments.experiment_1.train import compute_losses as compute_losses_exp1
from experiments.experiment_1.train import make_generate_epoch_data as make_gen_exp1
from experiments.experiment_1.train import make_validation_fn as make_val_exp1

from experiments.experiment_2.train import compute_losses as compute_losses_exp2
from experiments.experiment_2.train import make_generate_epoch_data as make_gen_exp2
from experiments.experiment_2.train import make_validation_fn as make_val_exp2
```

(These factories come from #88.)

### 3. Keeps only HPO-specific logic

```python
def run_training_trial(trial: optuna.trial.Trial, trial_cfg: FrozenDict) -> float:
    """Run a single HPO trial. Returns best NSE as float."""
    scenario = trial_cfg.get("scenario", "experiment_1")
    
    # 1. Init model (lightweight — no output dirs)
    key = random.PRNGKey(trial_cfg["training"]["seed"])
    model_key, train_key, val_key = random.split(key, 3)
    model, params = init_model_from_config(trial_cfg)  # or equivalent

    # 2. Create optimizer (same as production)
    num_batches = calculate_num_batches(batch_size, sample_sizes, ...)
    optimiser = create_optimizer(trial_cfg, num_batches=num_batches)
    opt_state = optimiser.init(params)

    # 3. Setup losses and weights (same as production)
    static_weights, _ = extract_loss_weights(trial_cfg)
    current_weights = get_active_loss_weights(static_weights, data_free=data_free, ...)
    
    # 4. Get experiment-specific functions
    compute_losses_fn = EXPERIMENT_REGISTRY[scenario]["compute_losses"]
    generate_epoch_data_jit = EXPERIMENT_REGISTRY[scenario]["make_gen"](...)
    validation_fn = EXPERIMENT_REGISTRY[scenario]["make_val"](...)

    # 5. Create scan body (same as production)
    scan_body = make_scan_body(
        train_step_jitted, model, optimiser, current_weights,
        trial_cfg, data_free, compute_losses_fn=compute_losses_fn,
    )

    # 6. Epoch loop with Optuna integration
    best_nse = -jnp.inf
    for epoch in range(epochs):
        train_key, epoch_key = random.split(train_key)
        scan_inputs = generate_epoch_data_jit(epoch_key)
        (params, opt_state), _ = lax.scan(scan_body, (params, opt_state), scan_inputs)

        # Validate + prune
        val_metrics = validation_fn(model, params)
        current_nse = val_metrics.get("nse_h", -jnp.inf)
        
        if jnp.isnan(current_nse):
            raise optuna.exceptions.TrialPruned()
        
        best_nse = max(best_nse, current_nse)
        trial.report(best_nse, epoch)
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    return float(best_nse) if best_nse > -jnp.inf else -1.0
```

### Key Design Decisions

1. **No checkpointing** — HPO trials only return a scalar; best configs are extracted post-hoc from Optuna DB
2. **No Aim logging** — no experiment tracking overhead during HPO
3. **No plotting** — no visualization during HPO
4. **No interactive prompts** — HPO runs unattended
5. **Same optimizer/loss/sampling as production** — ensures HP transfer fidelity

## Files to Modify

- `optimisation/optimization_train_loop.py` — **full rewrite**

## Files to Read (reference for the rewrite)

- `src/training/step.py` — `train_step_jitted` signature and usage
- `src/training/epoch.py` — `make_scan_body`, `sample_and_batch`, `empty_batch`, `maybe_batch_data`
- `src/training/optimizer.py` — `create_optimizer` signature
- `src/training/setup.py` — `extract_loss_weights`, `get_active_loss_weights`, `calculate_num_batches`, `get_sampling_count_from_config`, `get_boundary_segment_count`
- `experiments/experiment_1/train.py` — `compute_losses`, factory functions (after #88)
- `experiments/experiment_2/train.py` — `compute_losses`, factory functions (after #88)

## Acceptance Criteria

- [ ] `optimization_train_loop.py` is ≤150 lines
- [ ] No `train_step_trial()` — uses `train_step_jitted` + `make_scan_body()` from `src/training/`
- [ ] No `get_batches_tensor` or `get_sample_count` imports
- [ ] Uses `create_optimizer(cfg, num_batches)` — same optimizer as production
- [ ] Uses `extract_loss_weights()` + `get_active_loss_weights()` — same weight handling
- [ ] Uses `calculate_num_batches()` — same batch calculation
- [ ] Applies `_apply_min_depth()` during validation
- [ ] Imports `compute_losses` from experiment modules (not inlined)
- [ ] `python -m unittest discover test` passes
- [ ] Smoke test: run 2-3 Optuna trials with a small config and verify valid NSE returned
